hcl2template/types.packer_config.go

```diff
@@ -639,6 +642,47 @@ func (cfg *PackerConfig) getCoreBuildPostProcessors(source SourceUseBlock, block
 	return res, diags
 }
 
+// GetHCPPackerRegistryBlock return the HCP registry configuration block
+// that can should be used for the current build. Right now, it should
+// use the block at the top level but support the block inside the first
+// build block with a deprecation diagnostic
+func (cfg *PackerConfig) GetHCPPackerRegistry() (*HCPPackerRegistryBlock, hcl.Diagnostics) {
+	var block *HCPPackerRegistryBlock
+	var diags hcl.Diagnostics
+
+	multipleRegistryDiag := func(*HCPPackerRegistryBlock) *hcl.Diagnostic {
+		return &hcl.Diagnostic{
+			Summary:  "Multiple HCP Packer registry block declaration",
+			Subject:  block.HCL2Ref.DefRange.Ptr(),
+			Severity: hcl.DiagError,
+			Detail:   "Multiple HCP Packer Registry block has been found, please configure it at the top level",
+		}
+	}
+	// We start by looking in the build blcoks
+	for _, build := range cfg.Builds {
+		if build.HCPPackerRegistry != nil && block != nil {
+			// error multiple build block
+			diags = append(diags, multipleRegistryDiag(build.HCPPackerRegistry))
+			return nil, diags
+		}
+		if build.HCPPackerRegistry != nil {
+			block = build.HCPPackerRegistry
+		}
+	}
+
+	// then we look for the top level block
+	if block != nil && cfg.HCPPackerRegistry != nil {
+		diags = append(diags, multipleRegistryDiag(block))
+		return nil, diags
+	}
+
+	if block != nil {
```

Comment:
That should not be necessary I think, you can always return `block, diags`, if there's no block detected it will be nil, and so will the diags.
If there are diags here, especially errors, the caller should probably stop evaluation at this point, so I would think it's not problematic to return both at all times.